#! /bin/sh /usr/share/dpatch/dpatch-run
## 13_secu_0.99.6_bug-1582.dpatch by  <fpeters@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Backport of r21665 to 0.99.4

@DPATCH@
diff -urNad wireshark-0.99.4~/epan/dissectors/packet-ssl.c wireshark-0.99.4/epan/dissectors/packet-ssl.c
--- wireshark-0.99.4~/epan/dissectors/packet-ssl.c	2006-10-31 18:59:07.000000000 +0100
+++ wireshark-0.99.4/epan/dissectors/packet-ssl.c	2007-06-23 12:36:15.526330192 +0200
@@ -109,6 +109,7 @@
 #include <epan/dissectors/packet-x509af.h>
 #include <epan/emem.h>
 #include <epan/tap.h>
+#include <epan/expert.h>
 #include "packet-ssl.h"
 #include "packet-ssl-utils.h"
 
@@ -430,7 +431,7 @@
                                    SslDecryptSession *conv_data, guint8 content_type);
 
 
-static void dissect_ssl3_hnd_cli_hello(tvbuff_t *tvb,
+static void dissect_ssl3_hnd_cli_hello(tvbuff_t *tvb, packet_info *pinfo,
                                        proto_tree *tree,
                                        guint32 offset, guint32 length,
                                        SslDecryptSession* ssl);
@@ -1394,7 +1395,7 @@
                 break;
 
             case SSL_HND_CLIENT_HELLO:
-                dissect_ssl3_hnd_cli_hello(tvb, ssl_hand_tree, offset, length, ssl);
+                dissect_ssl3_hnd_cli_hello(tvb, pinfo, ssl_hand_tree, offset, length, ssl);
             break;
 
             case SSL_HND_SERVER_HELLO:
@@ -1636,7 +1637,7 @@
 }
 
 static void
-dissect_ssl3_hnd_cli_hello(tvbuff_t *tvb,
+dissect_ssl3_hnd_cli_hello(tvbuff_t *tvb, packet_info *pinfo,
        proto_tree *tree, guint32 offset, guint32 length,
        SslDecryptSession*ssl)
 {
@@ -1652,10 +1653,11 @@
      */
     proto_tree *ti;
     proto_tree *cs_tree;
-    guint16 cipher_suite_length;
+    gint cipher_suite_length;
     guint8  compression_methods_length;
     guint8  compression_method;
     guint16 start_offset;
+
     cipher_suite_length = 0;
     compression_methods_length = 0;
     start_offset = offset;
@@ -1685,9 +1687,17 @@
             ti = proto_tree_add_none_format(tree,
                                             hf_ssl_handshake_cipher_suites,
                                             tvb, offset, cipher_suite_length,
-                                            "Cipher Suites (%u suite%s)",
+                                            "Cipher Suites (%d suite%s)",
                                             cipher_suite_length / 2,
                                             plurality(cipher_suite_length/2, "", "s"));
+            if (cipher_suite_length % 2) {
+                proto_tree_add_text(tree, tvb, offset, 2,
+                    "Invalid cipher suite length: %d", cipher_suite_length);
+                expert_add_info_format(pinfo, NULL, PI_MALFORMED, PI_ERROR,
+                    "Cipher suite length (%d) must be a multiple of 2",
+                    cipher_suite_length);
+                return;
+            }
 
             /* make this a subtree */
             cs_tree = proto_item_add_subtree(ti, ett_ssl_cipher_suites);
